modules/livestore/partition_reader.go

```diff
@@ -0,0 +1,259 @@
+
+type PartitionReader struct {
+	services.Service
+
+	partitionID   int32
+	consumerGroup string
+	topic         string
+
+	client *kgo.Client
+	adm    *kadm.Client
+
+	consume consumeFn
+	metrics partitionReaderMetrics
+
+	logger log.Logger
+}
+
+func NewPartitionReaderForPusher(client *kgo.Client, partitionID int32, cfg ingest.KafkaConfig, consume consumeFn, logger log.Logger, reg prometheus.Registerer) (*PartitionReader, error) {
+	metrics := newPartitionReaderMetrics(partitionID, reg)
+	return newPartitionReader(client, partitionID, cfg, consume, logger, metrics)
+}
+
+func newPartitionReader(client *kgo.Client, partitionID int32, cfg ingest.KafkaConfig, consume consumeFn, logger log.Logger, metrics partitionReaderMetrics) (*PartitionReader, error) {
+	r := &PartitionReader{
+		partitionID:   partitionID,
+		consumerGroup: cfg.ConsumerGroup,
+		topic:         cfg.Topic,
+		client:        client,
+		adm:           kadm.NewClient(client),
+		consume:       consume,
+		metrics:       metrics,
+		logger:        log.With(logger, "partition", partitionID),
+	}
+
+	r.Service = services.NewBasicService(r.start, r.running, r.stop)
+	return r, nil
+}
+
+func (r *PartitionReader) start(context.Context) error {
+	return nil
+}
+
+func (r *PartitionReader) running(ctx context.Context) error {
+	offset, err := r.fetchLastCommittedOffsetWithRetries(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch last committed offset: %w", err)
+	}
+
+	r.client.AddConsumePartitions(map[string]map[int32]kgo.Offset{r.topic: {r.partitionID: offset}})
+	defer r.client.RemoveConsumePartitions(map[string][]int32{r.topic: {r.partitionID}})
+
+	for ctx.Err() == nil {
+		fetches := r.client.PollFetches(ctx)
+		if fetches.Err() != nil {
+			if errors.Is(fetches.Err(), context.Canceled) {
+				return nil
+			}
+			err := collectFetchErrs(fetches)
+			level.Error(r.logger).Log("msg", "encountered error while fetching", "err", err)
+			continue
+		}
+
+		r.recordFetchesMetrics(fetches)
+		if offset := r.consumeFetches(ctx, fetches); offset != nil {
+			r.commitOffset(ctx, *offset)
+		}
+	}
+
+	return nil
+}
+
+func (r *PartitionReader) stop(error) error {
+	level.Info(r.logger).Log("msg", "stopping partition reader")
+
+	r.client.Close()
+
+	return nil
+}
+
+func collectFetchErrs(fetches kgo.Fetches) (_ error) {
+	mErr := multierror.New()
+	fetches.EachError(func(_ string, _ int32, err error) {
+		// kgo advises to "restart" the kafka client if the returned error is a kerr.Error.
+		// Recreating the client would cause duplicate metrics registration, so we don't do it for now.
+		mErr.Add(err)
+	})
+	return mErr.Err()
+}
+
+func (r *PartitionReader) consumeFetches(ctx context.Context, fetches kgo.Fetches) *kadm.Offset {
+	records := make([]record, 0, fetches.NumRecords())
+
+	var lastRecord *kgo.Record
+
+	var (
+		minOffset = int64(math.MaxInt64)
+		maxOffset = int64(0)
+	)
+	fetches.EachRecord(func(rec *kgo.Record) {
```

Comment:
The slice of kgo.Records are directly available via i.e. `fetches[0].Topics[0].Partitions[0].Records`.  Is it important to collect all records from all partitions in 1 giant slice?  Since this will always only consume 1 partition, we can probably skip the new slice alloc overhead.

Reply:
this is a good catch, but i'm going to merge as is and we can improve perf later. it's on my list!